browser/ui/views/sidebar/sidebar_container_view.cc

```diff
@@ -348,43 +402,178 @@ SidebarContainerView::GetEventDetectWidget() {
   }
+
+  // Animation will trigger layout by changing preferred size.
+  if (ShouldUseAnimation()) {
+    DVLOG(1) << __func__ << ": show with animation";
+    width_animation_.Show();
+    return;
+  }
+
+  DVLOG(1) << __func__ << ": show w/o animation";
+  // Otherwise, layout should be requested here.
+  side_panel_->SetVisible(true);
+  InvalidateLayout();
 }
 
-void SidebarContainerView::DoHideSidebar(bool show_event_detect_widget) {
-  sidebar_control_view_->SetVisible(false);
-  ShowOptionsEventDetectWidget(show_event_detect_widget);
+void SidebarContainerView::HideSidebar(bool hide_sidebar_control) {
+  DVLOG(1) << __func__ << ": hide control: " << hide_sidebar_control;
+
+  if (width_animation_.is_animating()) {
+    DVLOG(1) << __func__ << ": Finish previous show/hide animation.";
+    width_animation_.End();
+  }
+
+  width_animation_.Reset(1.0);
+
+  // Calculate the start & end width for animation. Both are used when
+  // calculating preferred width during the hide animation.
+  animation_start_width_ = 0;
+  animation_end_width_ = 0;
+
+  if (sidebar_control_view_->GetVisible()) {
+    animation_start_width_ = sidebar_control_view_->GetPreferredSize().width();
+  }
+
+  if (side_panel_->GetVisible()) {
+    animation_start_width_ += side_panel_->GetPreferredSize().width();
+  }
+
+  if (!hide_sidebar_control) {
+    animation_end_width_ = sidebar_control_view_->GetPreferredSize().width();
+  }
+
+  if (animation_start_width_ == animation_end_width_) {
+    DVLOG(1) << __func__ << ": don't need hide animation.";
+
+    // At startup, make event detect widget visible even if children's
+    // visibility state is not changed.
+    if (animation_end_width_ == 0) {
+      ShowOptionsEventDetectWidget(true);
+    }
+
+    return;
+  }
+
+  DVLOG(1) << __func__ << ": hide animation (start, end) width: ("
+           << animation_start_width_ << ", " << animation_end_width_ << ")";
+
+  GetFocusManager()->ClearFocus();
+
+  if (ShouldUseAnimation()) {
+    DVLOG(1) << __func__ << ": hide with animation";
+    // Hide panel before starting animation to prevent contents layout during
+    // the animation.
+    side_panel_->SetVisible(false);
+    width_animation_.Hide();
+    return;
+  }
+
+  DVLOG(1) << __func__ << ": hide w/o animation";
+  if (animation_end_width_ == 0) {
+    ShowOptionsEventDetectWidget(true);
+  }
+
+  sidebar_control_view_->SetVisible(!hide_sidebar_control);
+  side_panel_->SetVisible(false);
   InvalidateLayout();
 }
 
+void SidebarContainerView::HideSidebarForShowOption() {
+  if (show_sidebar_option_ == ShowSidebarOption::kShowAlways) {
+    HideSidebar(false);
+  }
+
+  if (show_sidebar_option_ == ShowSidebarOption::kShowOnMouseOver) {
+    // Hide all if mouse is outside of control view.
+    HideSidebar(!sidebar_control_view_->IsMouseHovered());
+  }
+
+  if (show_sidebar_option_ == ShowSidebarOption::kShowNever) {
+    HideSidebar(true);
+  }
+}
+
+bool SidebarContainerView::ShouldUseAnimation() const {
```

Comment:
nit: It'd be good to consider gfx::Animation::ShouldRenderRichAnimation() together.

Reply:
This method checks `gfx::Animation::ShouldRenderRichAnimation()` already :)